File: WayforkWindows/app/lib/core/plan/host_resolver.dart
import 'dart:io';

import 'package:wayfork/core/model/store.dart';
import 'package:wayfork/core/singbox/constants.dart';
import 'package:wayfork/core/support/ipv4_prefix.dart';

/// Resolves OpenVPN server names for direct route rules. DNS lookup is
/// blocking below the Dart API, so callers should keep it off latency-sensitive
/// UI paths.
abstract final class HostResolver {
  /// Non-literal remotes of enabled OpenVPN tunnels, lowercase and unique in
  /// store order.
  static List<String> openVPNHosts(Store store) {
    final hosts = <String>[];
    for (final tunnel in store.tunnels.where((tunnel) => tunnel.isEnabled)) {
      final meta = tunnel.kind.openVPN;
      if (meta == null) continue;
      for (final remote in meta.remotes) {
        if (IPv4Prefix.parse(remote.host) != null) continue;
        final host = remote.host.toLowerCase();
        if (host.isNotEmpty && !hosts.contains(host)) hosts.add(host);
      }
    }
    return hosts;
  }

  /// IPv4 addresses per host, sorted and unique. Failed hosts are omitted and
  /// lookup errors never escape.
  static Future<Map<String, List<String>>> resolveIPv4(
    Iterable<String> hosts,
  ) async {
    final result = <String, List<String>>{};
    final fakeRange = IPv4Prefix.parse(SingBoxConstants.fakeIPv4Range)!;
    for (final host in hosts) {
      try {
        final addresses =
            (await InternetAddress.lookup(host, type: InternetAddressType.IPv4))
                .map((address) => address.address)
                .where((address) {
                  final prefix = IPv4Prefix.parse(address);
                  // A lookup while Wayfork is already On may receive Wayfork's
                  // own fake-IP answer, not a usable OpenVPN server address.
                  return prefix != null && !fakeRange.contains(prefix);
                })
                .toSet()
                .toList()
              ..sort();
        if (addresses.isNotEmpty) result[host] = addresses;
      } on Object {
        // Resolution is best-effort; the generator still matches the hostname.
      }
    }
    return result;
  }
}

File: macos/WayforkCore/Sources/WayforkCore/Plan/HostResolver.swift
import Foundation

/// Resolves the OpenVPN servers' hostnames for the plan (docs/design/03-routing.md, "Notes on
/// specific choices"). Blocking `getaddrinfo`: call it off the main actor.
public enum HostResolver {
    /// Non-literal OpenVPN and WireGuard server hosts, lowercased, unique, in store order.
    public static func serverHosts(in store: Store) -> [String] {
        var hosts: [String] = []
        for tunnel in store.tunnels where tunnel.isEnabled {
            let candidates: [String]
            switch tunnel.kind {
            case .openVPN(let meta): candidates = meta.remotes.map(\.host)
            case .wireGuard(let meta): candidates = meta.peers.map(\.host)
            case .vless, .shadowsocks, .trojan, .vmess: candidates = []
            }
            for candidate in candidates
            where candidate.contains("/") || IPv4Prefix(candidate)?.isHost != true {
                let host = candidate.lowercased()
                if !host.isEmpty, !hosts.contains(host) { hosts.append(host) }
            }
        }
        return hosts
    }

    /// Non-literal `remote` hosts of the enabled OpenVPN tunnels, lowercased, unique, in
    /// store order.
    public static func openVPNHosts(in store: Store) -> [String] {
        var hosts: [String] = []
        for tunnel in store.tunnels where tunnel.isEnabled {
            guard case .openVPN(let meta) = tunnel.kind else { continue }
            for remote in meta.remotes where IPv4Prefix(remote.host) == nil {
                let host = remote.host.lowercased()
                if !host.isEmpty, !hosts.contains(host) { hosts.append(host) }
            }
        }
        return hosts
    }

    /// IPv4 addresses per host (sorted, unique). Hosts that do not resolve are left out.
    public static func resolveIPv4(_ hosts: [String]) -> [String: [String]] {
        var result: [String: [String]] = [:]
        for host in hosts {
            // A lookup made while Wayfork is already On goes through Wayfork's own resolver
            // (F12), which answers with a fake IP for every name that is not yet in the
            // config's "server hosts → dns-direct" rule — the case on the first apply after
            // a tunnel is added. Such an answer is a name in disguise, never a server
            // address: pinning it as a WireGuard peer would dial the TUN in circles. The
            // Dart client has always filtered them; this is the Swift twin.
            let addresses = resolveIPv4(host).filter { !FakeIPIndex.isFakeIP($0) }
            if !addresses.isEmpty { result[host] = addresses }
        }
        return result
    }

    static func resolveIPv4(_ host: String) -> [String] {
        var hints = addrinfo()
        hints.ai_family = AF_INET
        hints.ai_socktype = SOCK_DGRAM
        var list: UnsafeMutablePointer<addrinfo>?
        guard getaddrinfo(host, nil, &hints, &list) == 0, let first = list else { return [] }
        defer { freeaddrinfo(first) }
        var addresses = Set<String>()
        var node: UnsafeMutablePointer<addrinfo>? = first
        while let entry = node {
            if entry.pointee.ai_family == AF_INET, let address = entry.pointee.ai_addr {
                address.withMemoryRebound(to: sockaddr_in.self, capacity: 1) { sin in
                    var addr = sin.pointee.sin_addr
                    var buffer = [CChar](repeating: 0, count: Int(INET_ADDRSTRLEN))
                    if inet_ntop(AF_INET, &addr, &buffer, socklen_t(INET_ADDRSTRLEN)) != nil {
                        addresses.insert(
                            buffer.withUnsafeBufferPointer { String(cString: $0.baseAddress!) })
                    }
                }
            }
            node = entry.pointee.ai_next
        }
        return addresses.sorted()
    }
}
